skbio/metadata/_interval.py

```diff
@@ -0,0 +1,669 @@
+            # Interval will take care of this
+            intvls = list(map(lambda x: (length-x[1], length-x[0]),
+                              reversed(f.locations)))
+            f.locations = intvls
+
+        # DONT' forget this!!!
+        self._is_stale_tree = True
+
+    @rebuild_tree
+    def sort(self, ascending=True):
+        '''Sort intervals by their starting and ending coordinates.
+
+        An interval with [(1, 2), (4, 7)] will be sorted in front
+        of another one with [(1, 2), (3, 8)].'''
+        self._intervals = sorted(
+            self._intervals,
+            key=lambda i: [i.locations[0][0], i.locations[-1][1]],
+            reverse=not ascending)
+
+    @rebuild_tree
+    def add(self, locations, boundaries=None, metadata=None):
+        """Adds a feature to the metadata object.
+
+        This method creates a ``Interval`` object and insert it into
+        the ``IntervalMetadata`` object.
+
+        Parameters
+        ----------
+        locations : iterable of tuple of ints
+            A list of locations associated with the Interval object.
+        boundaries : iterable of tuple of bool
+            A list of boundaries associated with the locations.
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            Interval object.
+
+        Returns
+        -------
+        Interval
+            The ``Interval`` object just added.
+
+        See Also
+        --------
+        skbio.metadata._add
+        """
+        # Add an interval to the tree. Note that the add functionality is
+        # built within the Interval constructor.
+        return Interval(interval_metadata=self,
+                        locations=locations,
+                        boundaries=boundaries,
+                        metadata=metadata)
+
+    @rebuild_tree
+    def _query_interval(self, location):
+        """Fetches Interval objects that overlap with the location."""
+        _assert_valid_location(location)
+
+        start, end = location
+        intvls = self._interval_tree.find(start, end)
+        # if a ``Interval`` has multiple non contiguous spans,
+        # multiple of which overlap with the location, then
+        # this ``Interval`` object will be returned
+        # multiple times. So we need to remove duplicates.
+        seen = set()
+        for intvl in intvls:
+            if id(intvl) not in seen:
+                seen.add(id(intvl))
+                yield intvl
+
+    @rebuild_tree
+    def _query_attribute(self, metadata, intervals=None):
+        """Fetches Interval objects based on query attributes.
+
+        Parameters
+        ----------
+        metadata : dict or ``None``
+            If it is ``None``, return empty iterator; if it is
+            ``{}``, return an interator of all the ``Interval``
+            objects.
+        intervals : a iterable of ``Interval`` objects
+        """
+        if metadata is None:
+            return
```

Comment:
it's valid to have a `return` and a `yield`? I'm a bit confused by the implication as it means this is both a generator and a function. Is it used as a short cut mechanism to avoid evaluation of the `metadata is not None` case? if so, wouldn't it be easier to do:

```python
if metadata is None:
    yield None
else:
    # do all the interesting stuff
```